tests/test_client_ws_functional.py

```diff
@@ -685,6 +685,35 @@ async def handler(request):
     assert resp.close_code is WSCloseCode.ABNORMAL_CLOSURE
 
 
+async def test_heartbeat_no_pong_concurrent_receive(aiohttp_client: Any) -> None:
+    ping_received = False
+
+    async def handler(request):
+        nonlocal ping_received
+        ws = web.WebSocketResponse(autoping=False)
+        await ws.prepare(request)
+        msg = await ws.receive()
+        if msg.type is aiohttp.WSMsgType.PING:
+            ping_received = True
```

Comment:
It looks like we never get anything but a PING here so maybe make this an assertion instead?

Reply:
The assertion is in the main test function. Can't really assert out of the request handler. The request will fail but it won't appear as a failed assertion.

Reply:
Do you have a suggestion for an alternate solution to solve the missing branch coverage?

Reply:
You mean the partial missing branch coverage in a test source code? 

I don't use coverage to cover the tests and in 20+ years haven't seen other people doing it either. The goal is to have 100% coverage of the code **under test**, not the test itself. It's like taking the EKG of the patient AND the attending physician. You *can* do it, but the EKG of the physician provides no actionable information.

I would simply exclude the entire test suite from coverage reporting.

Reply:
I'll leave this PR for other members to comment as how the test coverage reporting was setup before my time, and I don't know the history here.

Reply:
> I don't use coverage to cover the tests and in 20+ years haven't seen other people doing it either. The goal is to have 100% coverage of the code **under test**, not the test itself. It's like taking the EKG of the patient AND the attending physician. You _can_ do it, but the EKG of the physician provides no actionable information.

On the contrary, if test code isn't running (i.e. there isn't coverage) it shouldn't exist. When we added coverage for tests we found about a dozen tests that were not running at all, or not running the intended test code.

Reply:
> The assertion is in the main test function. Can't really assert out of the request handler. The request will fail but it won't appear as a failed assertion.

I think an assert is fine. The test will fail due to it not running the rest of the code, and the logs should show the assertion failure with a traceback.

Reply:
> On the contrary, if test code isn't running (i.e. there isn't coverage) it shouldn't exist. When we added coverage for tests we found about a dozen tests that were not running at all, or not running the intended test code.

The benefit of knowing whether the test code is running (it's the only arguable benefit, as you can't even correlate the coverage of the test code with coverage of the code under test) for me is absolutely not worth the effort required to achieve 100% line *and* branch coverage in the test. At worst you have some non-running code in your tests which is *functionally irrelevant* if you have 100%  line *and* branch coverage of the code under test.

Reply:
> I think an assert is fine. The test will fail due to it not running the rest of the code, and the logs should show the assertion failure with a traceback.

No it really isn't fine: 

```python
async def test_heartbeat_no_pong_concurrent_receive(aiohttp_client: Any) -> None:
    ping_received = False

    async def handler(request):
        nonlocal ping_received
        ws = web.WebSocketResponse(autoping=False)
        await ws.prepare(request)
        msg = await ws.receive()
        assert False
        if msg.type is aiohttp.WSMsgType.PING:
            ping_received = True
        ws._reader.feed_eof = lambda: None
        await asyncio.sleep(10.0)

    app = web.Application()
    app.router.add_route("GET", "/", handler)

    client = await aiohttp_client(app)
    resp = await client.ws_connect("/", heartbeat=0.1)
    resp._reader.feed_eof = lambda: None

    # Connection should be closed roughly after 1.5x heartbeat.
    msg = await resp.receive(5.0)
    # assert ping_received
    assert resp.close_code is WSCloseCode.ABNORMAL_CLOSURE
    assert msg
    assert msg.type is WSMsgType.ERROR
    assert isinstance(msg.data, ServerTimeoutError)
    
```

Results in a 100% passing test: 

```
test_client_ws_functional.py::test_heartbeat_no_pong_concurrent_receive[pyloop] /home/arcivanov/Documents/src/arcivanov/aiohttp/venv/.pyenv/versions/lib64/python3.12/site-packages/coverage/inorout.py:503: CoverageWarning: Module tests/ was never imported. (module-not-imported)
  self.warn(f"Module {pkg} was never imported.", slug="module-not-imported")
```

Reply:
> At worst you have some non-running code in your tests which is _functionally irrelevant_ if you have 100% line _and_ branch coverage of the code under test.

First, we don't have 100% coverage and achieving 100% coverage on the entire codebase would take a significant amount of work. Second, we have tests which test different expected behaviours of the code, even if a single test produces "full coverage" of that code (many of our regression tests when fixing bugs do not increase coverage of the code base). So, being aware that several of those tests are not even running is very useful. We've found numerous cases across projects where coverage has highlighted tests not running, broken tests or old code that should be removed or cleaned up (fixtures no longer used anywhere etc.).

> Results in a 100% passing test:

And an assert doesn't? I don't see how that makes it not fine to write an assert, the code makes sense either way, an assert just makes it clear that we should never see another type in this test.

Reply:
> And an assert doesn't? I don't see how that makes it not fine to write an assert, the code makes sense either way, an assert just makes it clear that we should never see another type in this test.


I have added an `assert False` right above `if msg.type is aiohttp.WSMsgType.PING:`. The test passes, the assert in the handler is ignored.

Reply:
Oh, I see what you're saying now. As I said, the test would fail due to the outer assert, and then I'd expect to see the inner assert log a traceback. To see the logs on a successful test, I think you need to run with `pytest -s`.

Reply:
Yes, running that myself after putting the outer assert back in, the test fails and I see this in the output:
```
------------------------------------------------------------------------------------------------- Captured log call -------------------------------------------------------------------------------------------------
ERROR    aiohttp.server:web_protocol.py:442 Error handling request
Traceback (most recent call last):
  File "/home/ubuntu/hacking/aiohttp/aiohttp/web_protocol.py", line 472, in _handle_request
    resp = await request_handler(request)
  File "/home/ubuntu/hacking/aiohttp/aiohttp/web_app.py", line 388, in _handle
    resp = await handler(request)
  File "/home/ubuntu/hacking/aiohttp/tests/test_client_ws_functional.py", line 951, in handler
    assert False
AssertionError: assert False
```

That's what I meant.

Reply:
So, do I understand you correctly: you're proposing replacing a functional assertion that works (works == fails the test when evaluated to False), with the one that doesn't work (i.e. will not fail the test) but will *indicate* a failure in the logs but only if the test harness is run in a verbose mode and only if manually parsed by human eye, all in pursuit of 100% test branch coverage which has no functional relevance for the code under test?

Reply:
And again, if the test gets messed up in future and ends up something like what you posted, then coverage will tell us that a chunk of our test is not running and there is probably an issue with the test. So, I think you just demonstrated a perfect example of why we do coverage here (and I vaguely recall such tests like this where the handlers were failing and the tests were still passing).

Reply:
> you're proposing replacing a functional assertion

No, I explicitly said that assertion needs to stay, not be replaced. I'm suggesting replacing an if statement which is unneeded (and suggests that could be a case where we get a different response) with an assert.

Reply:
> with the one that doesn't work

So, all we'd be doing is _adding_ information. The test would still fail on the same assertion, but we'd _also_ see a log showing the assertion failure in the handler, making it even easier for a developer to understand what went wrong when the test fails.

Reply:
```suggestion
        assert msg.type is aiohttp.WSMsgType.PING
        ping_received = True
```

Reply:
I have pushed a rewrite that eliminates the branch. Unfortunately I can't dedicate more time to this unless it's something functionally significant.